### ISSUE TITLE
New atom invalidation logic is incorrect

### ISSUE THREAD

**Issue by samkline** [pmndrs/jotai#1412]
Hi, we recently upgraded Jotai in our app from 1.7.0 to 1.8.2. This was a great performance improvement for our app, but we've found a subtle bug that sometimes causes `useAtomValue` to return old/stale values. We're using React 17.0.2 without Suspense and can reproduce it on both our web app and React Native app.

In particular, we have many usages of `useAtomValue` for the same derived atom, and some instances return the correct, current value while other instances are "stuck" returning the old, stale value. 

Unfortunately I don’t have a simple reproduction, as our application is quite complex with a large tree of primitive and derived atoms, and we can only reproduce this when performing some complex operations that update many of those atoms.

However, I’ve noticed that when commenting out these lines: https://github.com/pmndrs/jotai/blob/fc96b2c72a057cb00afef0dfbe2ca30cbb7b23c4/src/core/store.ts#L866-L875 we get correct behavior, but we also get extra re-renders, and of course Jotai’s ‘no extra rerenders after commit with derived atoms (#1213)’ optimization test fails.

I’ll keep digging to see if I can identify the cause or a simple reproduction case, but wanted to raise this issue in case the cause is obvious to one of the maintainers.

Thanks!

**Comment by dai-shi** [comment]
Thanks for reporting.
The logic around there is something I'm not super confident.
Hope you can reproduce the issue at minimum and create a failing test.

**Comment by edongashi** [comment]
Something really strange is happening. Using `useAtomValue` from `jotai/utils` does not work (stuck at initial value), while `useAtomValue` from `jotai` works!

Repro:

```tsx
import React, { FunctionComponent } from 'react';
import { atom, useAtom } from 'jotai';
import { useAtomValue } from 'jotai/utils';

const dataAtom = atom(0);

export const Component: FunctionComponent = () => {
  const [data1, setData] = useAtom(dataAtom);
  const data2 = useAtomValue(dataAtom);

  return (
    <div>
      useAtom: {data1} <br />
      useAtomValue: {data2} <br />
      <button onClick={() => setData((x) => x + 1)}>Increment</button>
    </div>
  );
}
```

![image](https://user-images.githubusercontent.com/12145268/199858011-3ca6c65f-9c25-4295-9b18-f62d6872ac8a.png)

The same code as above, but `useAtomValue` imported from core `jotai`:

![image](https://user-images.githubusercontent.com/12145268/199858153-d6e33354-55a6-4c67-a039-c84ec1c35ab0.png)

Environment is electron forge with webpack/typescript.
Could the bundler somehow have pulled a different distribution, resulting in 2 separate instances of the library?

**Comment by dai-shi** [comment]
https://github.com/pmndrs/jotai/blob/2f06e8e5b64d69af1516d2ec14947d7fc6498316/src/utils.ts#L3

> Could the bundler somehow have pulled a different distribution, resulting in 2 separate instances of the library?

Probably.

**Comment by edongashi** [comment]
Testing indeed shows that the two `useAtomValue`s are not referentially equal.

Digging into the webpack dev bundle I see that it has emitted both `./node_modules/jotai/esm/index.js` and `./node_modules/jotai/esm/index.mjs`, so it must be referencing both in different contexts.

I thought I ran into the invalidation bug discussed in the issue, but this is unrelated...

Any risk in making the above a relative import?

**Comment by dai-shi** [comment]
It's known as dual package hazard, and we know this library suffers from it.
We spent lots of efforts previously, but at the moment, we think to leave the problem. (as we didn't find reasonable solutions that work well for us.)

Hope you can configure your env to only use one bundle. Would also love to see a minimal reproduction. Please open a new discussion.

**Comment by edongashi** [comment]
Thank you @dai-shi for the pointer. Posted a description and potential fix in #1537. Sorry for accidentally hijacking the thread.

**Comment by dai-shi** [comment]
@samkline Can you please try if this problem occurs with v2 API? https://github.com/pmndrs/jotai/blob/50a5e8ac87223e068f7704644fae90ab6103c0fa/docs/guides/migrating-to-v2-api.mdx

Please open a new issue if the problem persists.

Closing this as being a bit stale.
